### ISSUE TITLE
OpenTelemetry Support

### ISSUE THREAD

**Issue by JoJo2406** [containous/traefik#6374]
**Feature**

### OpenTelemetry Support

It seems OpenTracing and OpenCensus have merged to form OpenTelemetry (https://opentracing.io/)

Could we look to add support / migrate in a future release.

**Comment by maxisme** [comment]
Hey can I ask why you need OpenTelemetry support? As Traefik already has OpenTracing?

**Comment by dPeS** [comment]
@maxisme `In May 2019, the Cloud Native Computing Foundation announced the merging of the OpenTracing and OpenCensus projects into a new standard called OpenTelemetry.`

**Comment by dPeS** [comment]
+1 waiting for opentelemetry here

**Comment by maxisme** [comment]
@dPeS Yes, I was just wondering if there were any features that would be added when moving to OpenTelemetry. For my apps, I currently use `OpenTelemetry` along with `b3` (openzipkin) in Traefik. I am quite new to tracing but would it basically just enable my opentelemetry tracing to work seamlessly without having to specify `propagators.set_global_httptextformat(B3Format())` for example in python?

**Comment by mars64** [comment]
re: opentracing vs opentelemetry, I think the win is "improved ergonomics" -- I won't have to do things like support deprecated jaeger receivers. [Since the jaeger team has already declared `future Jaeger backend components will be based on OpenTelemetry collector`](https://www.jaegertracing.io/docs/1.21/opentelemetry/), it seems prudent to support opentelemetry.

re: additional features/capabilities in the future, opentelemetry also aims to be a [unified metrics and logging proxy](https://github.com/open-telemetry/opentelemetry-go#project-status), though those efforts are deprioritized for now.

I imagine that there won't be a ton of changes necessary for `instrumentation exporters` though, e.g. in your `propagator` example -- but I'm not certain there.

**Comment by Azuka** [comment]
Just adding that additionally, opentelemetry has a larger extended set of [semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/README.md).

**Comment by ryandens** [comment]
:wave: Would the Traefik team be interested in having a community member take this on?  I'd be happy to take a stab at it. If you are, what shape would you prefer it to take? I'm imagining that it makes sense to add support for OpenTelemetry alongside OpenCensus at this time, but I also understand that might be too high a maintenance burden and that you might want to wait for a time when it's prudent to simply cut over from OpenTracing to OpenTelemetry.

If you're willing to consider a change to introduce this now, I could see this taking one of three forms:
1. Cutover from OpenTracing to OpenTelemetry (I have a hunch that this would be too abrupt for Traefik users, but this is the lowest maintenance burden if you'd like to move forward with adding OpenTelemetry support)
2.  Add OpenTelemetry as another tracing `tracing.Backend` implementation. This would require some refactoring of the `pkg/tracing` API I think, or we could handle the complexity in some kind of shim between the OpenTracing API and the OpenTelemetry API. To users, this might present as the following configuration:

```yaml
tracing:
  opentelemetry:
    exporters:
      otlp:
        endpoint: localhost:4317
        protocol: grpc
```

Pros: more stable internal API for tracing, allows OpenTelemetry and OpenTracing to exist side-by-side for a long time
Cons:  If you want to remove OpenTracing support in the future, it could present as needlessly complex to users. In addition, if you wanted to add OpenTelemetry Metrics support in the future, it could be challenging to present a unified OpenTelemetry experience to Traefik users who want to leverage both.

3. Add OpenTelemetry as its own top-level Observability construct. This would likely mean a different tracing middleware for OpenTelemetry than OpenTracing, but we could still share a decent amount of logic between the two if we felt that was the right thing to do. This could present to users as the following configuration:

```yaml
opentelemetry:
  exporters:
    otlp:
      endpoint:  localhost:4317
      protocol: grpc
  tracingEnabled: true
  metricsEnabled: true
```

or like this:

```yaml
opentelemetry:
  tracing:
    exporters:
      otlp:
        endpoint:  localhost:4317
        protocol: grpc
  metrics:
    exporters:
      otlp:
        endpoint:  localhost:4317
        protocol: grpc
```

The metrics configurations wouldn't have to be added right away (since the development of the OpenTelemetry Metrics API is currently paused) but I included them in this example to illustrate what it might look like

Pros: A better long-term story for Traefik users to leverage a unified, open-source API for observability (tracing, metrics, and logs) while still allowing the existence of the old OpenTracing implementation in this project. 
Cons: Would probably be a bigger API change and could be a bigger maintenance burden, depending on the long-term goals of Traefik's observability posture.


If you don't think now is a  good time for this undertaking, I totally understand! :smiley:

**Comment by mrsufgi** [comment]
Seems that with OpenTelemetry v1.0 we can move forward with this, implementing #2 and making it the default(!) would be the way to go. 

In the meantime, since otel-collector supports jaeger (via a receiver) do you have any ideas for a viable workaround?
I thought changing the `uber-trace-id` would work, but probably misconfigured.

@ryandens are you planning to implementing this? maybe @PeterSzegedi (PR for lightstep closed due to being too specific) could help?

I would also be happy to help with this <3  
Really looking forward having full traefik->service-n(service)->db trace setup.

**Comment by ddtmachado** [comment]
Hey everyone, it's been quite some time and we didn't heard back from @ryandens or anyone else.

We really want to push this forward so we're taking over the task now and performing a few studies for the integration. Thanks again for the discussions and proposals and keep tuned for updates on the progress.

**Comment by bixu** [comment]
Hi @ddtmachado, my org is moving quickly to an all-Otel tracing strategy and are also evaluating Traefik at the moment. Is there anything we can do to support this effort? Traefik + Otel feels pretty compelling to us.

**Comment by ddtmachado** [comment]
Hi @bixu , thank you for the feedback! We also feel this feature is really compelling and something the community is really looking for. Having said that we'll probably delay this integration until the next major version of Traefik Proxy, 3.0 which we still plan on delivering before the end of the year.

This delay will allow us to focus on smaller improvements that are already work in progress and ready to be shipped on the next minor.

**Comment by bixu** [comment]
@ddtmachado, Otel support came up again recently at $job -- does your team have an early-access channel for upcoming releases? I think we might be open to working with you closely on EA testing, possibly even contributing.

**Comment by ddtmachado** [comment]
@bixu, that's great! we don't have an official docker image or binary publication for every PR, altought that is something we considered and maybe we should go ahead at some point, that means you'll need to build the PR to try it out, we have a guide on how to build Traefik [here](https://doc.traefik.io/traefik/contributing/building-testing/).

We're eager to get feedback on this new integration so please share your results!

**Comment by gurleensethi** [comment]
Any updates on OTEL tracing support?

**Comment by ddtmachado** [comment]
@gurleensethi it will be available on the next major version, 3.0, as everything with the `next` milestone tag will ;)
